src/webgpu/shader/execution/shader_io/compute_builtins.spec.ts

```diff
@@ -254,3 +255,424 @@ g.test('inputs')
+ * @param numWGs Number of workgroups
+ */
+function checkSubgroupInvocationIdConsistency(
+  data: Uint32Array,
+  subgroupSize: number,
+  wgSize: number,
+  numWGs: number
+): Error | undefined {
+  const partialSize = wgSize % subgroupSize;
+  const numFullSubgroups = Math.floor(wgSize / subgroupSize);
+
+  for (let wg = 0; wg < numWGs; wg++) {
+    const ids = [...iterRange(subgroupSize, x => 0)];
+    for (let i = 0; i < wgSize; i++) {
+      const idx = i + wg * wgSize;
+      const id = data[idx];
+      if (subgroupSize <= data[idx]) {
+        return new Error(`Subgroup invocation id '${id}' exceeded subgroup size '${subgroupSize}'`);
+      }
+      ids[id]++;
+    }
+    for (let i = 0; i < ids.length; i++) {
+      let expect = numFullSubgroups;
+      if (i < partialSize) {
+        expect++;
+      }
+      if (ids[i] !== expect) {
+        return new Error(
+          ErrorMsg(`Workgroup ${wg}: subgroup_invocation_id ${i} inconsistent`, ids[i], expect)
+        );
+      }
+    }
+  }
+
+  return undefined;
+}
+
+g.test('subgroup_invocation_id')
+  .desc(
+    'Tests subgroup_invocation_id values. No mapping between local_invocation_index and subgroup_invocation_id can be relied upon.'
+  )
+  .params(u =>
+    u
+      .combine('sizes', kWGSizes)
+      .beginSubcases()
+      .combine('numWGs', [1, 2] as const)
+  )
+  .beforeAllSubcases(t => {
+    t.selectDeviceOrSkipTestCase('subgroups' as GPUFeatureName);
+  })
+  .fn(async t => {
+    const wgThreads = t.params.sizes[0] * t.params.sizes[1] * t.params.sizes[2];
+
+    // Compatibility mode has lower workgroup limits.
+    const {
+      maxComputeInvocationsPerWorkgroup,
+      maxComputeWorkgroupSizeX,
+      maxComputeWorkgroupSizeY,
+    } = t.device.limits;
+    t.skipIf(
+      maxComputeInvocationsPerWorkgroup < wgThreads ||
+        maxComputeWorkgroupSizeX < t.params.sizes[0] ||
+        maxComputeWorkgroupSizeY < t.params.sizes[1],
+      'Workgroup size too large'
+    );
+
+    const wgsl = `
+enable subgroups;
+
+const stride = ${wgThreads};
+
+@group(0) @binding(0)
+var<storage, read_write> output : array<u32>;
+
+@group(0) @binding(1)
+var<storage, read_write> sizes : array<u32>;
+
+@compute @workgroup_size(${t.params.sizes[0]}, ${t.params.sizes[1]}, ${t.params.sizes[2]})
+fn main(@builtin(subgroup_size) size : u32,
+        @builtin(subgroup_invocation_id) id : u32,
+        @builtin(workgroup_id) wgid : vec3u,
+        @builtin(local_invocation_index) lid : u32) {
+  output[lid + stride * wgid.x] = id;
+  if (lid == 0) {
+    sizes[0] = size;
+  }
+}`;
+
+    const numInvocations = wgThreads * t.params.numWGs;
+    const outputBuffer = t.makeBufferWithContents(
+      new Uint32Array([...iterRange(numInvocations, x => 0)]),
+      GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST
+    );
+    t.trackForCleanup(outputBuffer);
+    const sizeBuffer = t.makeBufferWithContents(
+      new Uint32Array([0]),
```

Comment:
make this the number of workgroups (depending on how the shader avoids the race)